### DISCUSSION TITLE
How to addEventListener to onMounted hooks

### DISCUSSION THREAD

**Discussion by adriaandotcom** [nuxt/nuxt#16832] [Questions]
It's a bit of a mix between Nuxt 3 and Vue 3.

`pages/blog/[slug].vue`:

```vue
<template>
  <div v-html="article"></div>
</template>

<script>
const { pending, data: post } = await useAsyncData(
  `blog-${route.params.slug}`,
  () => $fetch("/api/blog-post?slug=" + route.params.slug)
);

const article = computed(() => {
  if (!post?.value?.article) return null;
  return post.value.article;
});

if (process.client) {
  onMounted(() => {
    const gifs = document.querySelectorAll("div.gif");
    if (!gifs.length) return;
    gifs.forEach((gif) => {
      gif.addEventListener("mouseenter", () => {
        gif.querySelector("details").open = true;
      });
      gif.addEventListener("mouseleave", () => {
        gif.querySelector("details").open = false;
      });
    });
  });
}
</script>
```

It needs to add an `addEventListener` to all GIFs on the page. The HTML comes from an API, so I need to add some sugar for the interactions. I need a _details_-tag to open when a mouse enters the _div_ with class `gif`.

It works fine on the first page load. But when navigating to another page the `addEventListener` are not applied.  However, the `onMounted` is called between navigations. It seems to me that they are applied to a (shadow) copy of the page instead of the actual HTML being shown to the visitors.

**Comment by adriaandotcom** [comment]
Funny enough, with a `setTimeout` it does work:

```js
if (process.client) {
  onMounted(() => {
    setTimeout(() => {
      const gifs = document.querySelectorAll("div.gif");
      if (!gifs.length) return;
      gifs.forEach((gif) => {
        gif.addEventListener("mouseenter", () => {
          gif.querySelector("details").open = true;
        });
        gif.addEventListener("mouseleave", () => {
          gif.querySelector("details").open = false;
        });
      });
    }, 1000);
  });
}
```

**Comment by TheAlexLichter** [comment]
1. You don't need `process.client` - `onMounted` is executed on the client side only anyway.
2. If the gifs are created by the `v-html` directive, you can use `setTimeout` or better, Vue's `nextTick` to wait so the elements are actually part of the DOM before executing your code

**Reply by adriaandotcom** [reply]
Using `nextTick` this way, does not work:

```js
onMounted(() => {
  nextTick(() => {
    const gifs = document.querySelectorAll("div.gif");
    if (!gifs.length) return;
    gifs.forEach((gif) => {
      gif.addEventListener("mouseenter", () => {
        gif.querySelector("details").open = true;
      });
      gif.addEventListener("mouseleave", () => {
        gif.querySelector("details").open = false;
      });
    });
  });
});
```

Thanks for the tip on `process.client`, removed that.

**Reply by adriaandotcom** [reply]
I added the `await useAsyncData` code in the top description as well.

**Reply by TheAlexLichter** [reply]
Can you try using `await nextTick` instead of using the callback? Shouldn't change much in theory but well 🙈 

You can test the existance of the gifs easily by logging. If `nextTick` isn't "enough", then there might be some kind of race condition.

Also, do you use SSR or not?

**Reply by adriaandotcom** [reply]
I use the hybrid version, so I guess that's SSR as well.
